src/HDF5.jl

```diff
@@ -718,6 +720,8 @@ function create_dataset(parent::Union{File,Group}, path::AbstractString, dtype::
     dcpl = isempty(pv) ? DEFAULT_PROPERTIES : create_property(H5P_DATASET_CREATE; pv...)
     dxpl = isempty(pv) ? DEFAULT_PROPERTIES : create_property(H5P_DATASET_XFER; pv...)
     dapl = isempty(pv) ? DEFAULT_PROPERTIES : create_property(H5P_DATASET_ACCESS; pv...)
+    checkvalid(parent)
+    haskey(parent, path) && error("cannot create dataset: object \"", path, "\" already exists at ", name(parent))
     Dataset(h5d_create(checkvalid(parent), path, dtype, dspace, _link_properties(path), dcpl, dapl), file(parent), dxpl)
```

Comment:
```suggestion
    checkvalid(parent)
    haskey(parent, path) && error("cannot create dataset: object \"", path, "\" already exists at ", name(parent))
    Dataset(h5d_create(parent, path, dtype, dspace, _link_properties(path), dcpl, dapl), file(parent), dxpl)
```

Though I would also suggest getting rid of the explicit `checkvalid(parent)` in both functions and to edit `haskey`'s implementation to do the `checkvalid` (which is right now inconsistently applied — checked for attributes, datasets, and datatypes but not for files and groups).

That could be left to a followup  PR if you prefer, though.